analyzers/src/SonarAnalyzer.CSharp/Rules/GenericTypeParameterUnused.cs

```diff
@@ -37,123 +38,82 @@ public sealed class GenericTypeParameterUnused : SonarDiagnosticAnalyzer
         private const string MessageFormat = "'{0}' is not used in the {1}.";
 
         private static readonly DiagnosticDescriptor Rule = DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        private static readonly SyntaxKind[] MethodModifiersToSkip =
+        {
+            SyntaxKind.AbstractKeyword,
+            SyntaxKind.VirtualKeyword,
+            SyntaxKind.OverrideKeyword
+        };
 
         public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(Rule);
 
-        protected override void Initialize(SonarAnalysisContext context) =>
-            context.RegisterCompilationStartAction(analysisContext =>
-            {
-                analysisContext.RegisterSyntaxNodeAction(c =>
-                     {
-                         var declarationSymbol = c.SemanticModel.GetDeclaredSymbol(c.Node);
-                         if (declarationSymbol == null)
-                         {
-                             return;
-                         }
-
-                         CheckGenericTypeParameters(declarationSymbol, c);
-                     },
-                     SyntaxKind.MethodDeclaration,
-                     SyntaxKindEx.LocalFunctionStatement);
-
-                analysisContext.RegisterSyntaxNodeAction(c =>
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(c =>
+                {
+                    if (c.SemanticModel.GetDeclaredSymbol(c.Node) is { } declarationSymbol)
+                    {
+                        CheckGenericTypeParameters(c, declarationSymbol);
+                    }
+                },
+                SyntaxKind.MethodDeclaration,
+                SyntaxKindEx.LocalFunctionStatement);
+
+            context.RegisterSyntaxNodeActionInNonGenerated(c =>
+                 {
+                     if (c.IsRedundantPositionalRecordContext())
                      {
-                         if (c.ContainingSymbol.Kind != SymbolKind.NamedType)
-                         {
-                             return;
-                         }
-
-                         CheckGenericTypeParameters(c.ContainingSymbol, c);
-                     },
-                     SyntaxKind.ClassDeclaration,
-                     SyntaxKindEx.RecordClassDeclaration);
-            });
+                         CheckGenericTypeParameters(c, c.ContainingSymbol);
+                     }
+                 },
+                 SyntaxKind.ClassDeclaration,
+                 SyntaxKindEx.RecordClassDeclaration,
+                 SyntaxKindEx.RecordStructDeclaration,
+                 SyntaxKind.StructDeclaration);
```

Comment:
Does it make sense to add interfaces too? An interface, that declares a type parameter, but it doesn't use it is at least a code smell IMO.